rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,453 @@
+
+4. **The Server must verify that the client requesting certificates is the only client that can use the certificates**
+
+Like other login procedures, this be accomplished with a PKCE-like flow, where the client provides their public key in the login request, and the server issues certificates for the client's public key. If the client does not have access to the corresponding client private key, they cannot use the certificates.
+
+Note: Ideally, we would ensure that the only client that can *receive* certificates is the requesting client, but ensuring the PKCE flow above is enough to defend against attacks since the certificates are useless without the client's private key. In the current design this is accomplished by encapsulating the headless login flow between the remote machine and Teleport into a single endpoint - `/webapi/login/headless`. Additionally, the certificates returned will be encrypted by the client's public key so it can only be read with the client's private key.
+
+5. **Limit the scope of headless certificates**
+
+When a user requests headless login certificates, we know exactly what command they are trying to complete. This means that we can scope their certificate to provide the most limited possible privileges necessary to complete said action. This would help to reduce the blast radius if an attacker manages to steal the user's headless certificates.
+
+However, unlike Per-session MFA, which scopes certificates to a specific node connection, we will need to scope the certificates to the entire API flow for a `tsh` command. For example, `tsh ssh` needs to:
+
+* Connect to Proxy
+* List nodes
+* View cluster details (Auth preference, etc.)
+* Connect to Node
+
+This means that for any given `tsh` command, the Teleport Auth server will provide different privileges to the certificates. Since each possible command will need custom logic to determine the lowest privileges necessary, we will only support a select few commands necessary for basic `tsh` workflows:
+
+* `tsh ls`
+* `tsh ssh`
+* `tsh scp`
+
+The awarded certificates should not provide access to roles usually awarded to the user, as these roles may provide more privileges than required for the requested command. However, some role provided fields should be included, such as `role.options.forward_agent` and `role.logins`. Additionally, the user should be notified of what command was requested and what permissions will be awarded to the headless session once approved.
+
+Example:
+
+```
+Incoming headless request for command "tsh ssh server01", to connect to node <server01_uuid>. Upon approval, the request will be awarded the following permissions:
```

Comment:
This will be done as an edge case similar to resource requests. I don't have a very specific answer here, as it is very in the weeds of implementation. It may be something like: 

The Server gets the auth request and pulls node resource IDs, etc. from the request. Then it combines this with the user's normal roles (excluding deny/allow rules) to form a limited user identity. We may need to encode a "headless" extension in the certificate so that the auth server knows to re-create this identity on proceeding requests.

Since this part of the design will not be included in the initial implementation, my plan is to follow up this unanswered questions in later implementation itself, at which point I will also update this RFD.